<!DOCTYPE html>
<html lang="en">
<head>
<meta charset="utf-8">
<meta name="viewport" content="width=device-width, initial-scale=1">
<title>Is there a difference between "throw" and "throw ex"? | StackOverflow Snapshot</title>
<link href="../../assets/css/all-nocdn.css" rel="stylesheet" type="text/css">
<meta name="theme-color" content="#5670d4">
<meta name="generator" content="Nikola (getnikola.com)">
<link rel="canonical" href="https://ohstackoverflow.netlify.app/posts/is-there-a-difference-between-throw-and-throw-ex/">
<!--[if lt IE 9]><script src="../../assets/js/html5.js"></script><![endif]--><meta name="author" content="Arya">
<link rel="prev" href="../how-to-uncheck-a-radio-button/" title="How to uncheck a radio button?" type="text/html">
<link rel="next" href="../how-can-i-post-data-as-form-data-instead-of-a-request-payload/" title="How can I post data as form data instead of a request payload?" type="text/html">
<meta property="og:site_name" content="StackOverflow Snapshot">
<meta property="og:title" content='Is there a difference between "throw" and "throw ex"?'>
<meta property="og:url" content="https://ohstackoverflow.netlify.app/posts/is-there-a-difference-between-throw-and-throw-ex/">
<meta property="og:description" content='There are some posts that asks what the difference between those two are
already.
(why do I have to even mention this...)
But my question is different in a way that I am calling "throw ex" in another
'>
<meta property="og:type" content="article">
<meta property="article:published_time" content="2023-02-18T06:16:50+08:00">
<meta property="article:tag" content=".net">
<meta property="article:tag" content=".net-c#">
<meta property="article:tag" content="exception">
</head>
<body>
<a href="#content" class="sr-only sr-only-focusable">Skip to main content</a>

<!-- Menubar -->

<nav class="navbar navbar-inverse navbar-static-top"><div class="container">
<!-- This keeps the margins nice -->
        <div class="navbar-header">
            <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#bs-navbar" aria-controls="bs-navbar" aria-expanded="false">
            <span class="sr-only">Toggle navigation</span>
            <span class="icon-bar"></span>
            <span class="icon-bar"></span>
            <span class="icon-bar"></span>
            </button>
            <a class="navbar-brand" href="https://ohstackoverflow.netlify.app/">

                <span id="blog-title">StackOverflow Snapshot</span>
            </a>
        </div>
<!-- /.navbar-header -->
        <div class="collapse navbar-collapse" id="bs-navbar" aria-expanded="false">
            <ul class="nav navbar-nav">
<li>
<a href="../../archive.html">Archive</a>
                </li>
<li>
<a href="../../categories/">Tags</a>

                
            </li>
</ul>
<ul class="nav navbar-nav navbar-right">
<div style="display:table;min-height:5rem;min-width:27rem;">
					<div class="input-group" style="display: table-cell;vertical-align: middle;">
						<input id="words" type="text" class="form-control" style="max-width:22rem;" onkeydown="if(event.keyCode==13){btn.click()}"><span class="input-group-btn" style="float:left">
							<button id="btn" class="btn btn-default" type="button" data-toggle="modal" data-target="#myModal">
								<span class="glyphicon glyphicon-search">
							</span></button>
						</span>
					</div>
<!-- /input-group -->
				</div>

				
                
                
                
            </ul>
</div>
<!-- /.navbar-collapse -->
    </div>
<!-- /.container -->
</nav><!-- End of Menubar --><!-- 模态框（Modal） --><div class="modal fade" id="myModal" tabindex="-1" role="dialog" aria-labelledby="myModalLabel" aria-hidden="true">
	<div class="modal-dialog">
		<div class="modal-content">
			<div class="modal-header">
				<button type="button" class="close" data-dismiss="modal" aria-hidden="true">×
				</button>
				<h4 class="modal-title" id="myModalLabel">
					查找结果
				</h4>
			</div>
			<div class="modal-body">
				<div id="search-count" style="min-height:4rem;">
				查找中，请稍后...
				</div>
				<div id="search-result">
				</div>

				
			</div>
			<div class="modal-footer">
				<button type="button" class="btn btn-default" data-dismiss="modal">
					关闭
				</button>
			</div>
		</div>
<!-- /.modal-content -->
	</div>
<!-- /.modal-dialog -->
</div>
<!-- /.modal -->

<div class="container" id="content" role="main">
    <div class="body-content">
        <!--Body content-->
        <div class="row">
            
            
<article class="post-text h-entry hentry postpage" itemscope="itemscope" itemtype="http://schema.org/Article"><header><h1 class="p-name entry-title" itemprop="headline name"><a href="." class="u-url">Is there a difference between "throw" and "throw ex"?</a></h1>

        <div class="metadata">
            <p class="byline author vcard p-author h-card"><span class="byline-name fn p-name" itemprop="author">
                    <a class="u-url" href="../../authors/arya/">Arya</a>
            </span></p>
            <p class="dateline">
            <a href="." rel="bookmark">
            <time class="published dt-published" datetime="2023-02-18T06:16:50+08:00" itemprop="datePublished" title="2023-02-18 06:16">2023-02-18 06:16</time></a>
            </p>
            

        </div>
        

    </header><div class="e-content entry-content" itemprop="articleBody text">
    <p>There are some posts that asks what the difference between those two are
already.<br>
(why do I have to even mention this...)</p>
<p>But my question is different in a way that I am calling "throw ex" in another
error <em>god-like</em> handling method.</p>
<div class="code"><pre class="code literal-block"><span class="nv">public</span><span class="w"> </span><span class="nv">class</span><span class="w"> </span><span class="nv">Program</span><span class="w"> </span>{
<span class="w">    </span><span class="nv">public</span><span class="w"> </span><span class="nv">static</span><span class="w"> </span><span class="nv">void</span><span class="w"> </span><span class="nv">Main</span><span class="ss">(</span><span class="nv">string</span>[]<span class="w"> </span><span class="nv">args</span><span class="ss">)</span><span class="w"> </span>{
<span class="w">        </span><span class="nv">try</span><span class="w"> </span>{
<span class="w">            </span><span class="o">//</span><span class="w"> </span><span class="nv">something</span>
<span class="w">        </span>}<span class="w"> </span><span class="nv">catch</span><span class="w"> </span><span class="ss">(</span><span class="nv">Exception</span><span class="w"> </span><span class="nv">ex</span><span class="ss">)</span><span class="w"> </span>{
<span class="w">            </span><span class="nv">HandleException</span><span class="ss">(</span><span class="nv">ex</span><span class="ss">)</span><span class="c1">;</span>
<span class="w">        </span>}
<span class="w">    </span>}

<span class="w">    </span><span class="nv">private</span><span class="w"> </span><span class="nv">static</span><span class="w"> </span><span class="nv">void</span><span class="w"> </span><span class="nv">HandleException</span><span class="ss">(</span><span class="nv">Exception</span><span class="w"> </span><span class="nv">ex</span><span class="ss">)</span><span class="w"> </span>{
<span class="w">        </span><span class="k">if</span><span class="w"> </span><span class="ss">(</span><span class="nv">ex</span><span class="w"> </span><span class="nv">is</span><span class="w"> </span><span class="nv">ThreadAbortException</span><span class="ss">)</span><span class="w"> </span>{
<span class="w">            </span><span class="o">//</span><span class="w"> </span><span class="nv">ignore</span><span class="w"> </span><span class="k">then</span>,
<span class="w">            </span><span class="k">return</span><span class="c1">;</span>
<span class="w">        </span>}
<span class="w">        </span><span class="k">if</span><span class="w"> </span><span class="ss">(</span><span class="nv">ex</span><span class="w"> </span><span class="nv">is</span><span class="w"> </span><span class="nv">ArgumentOutOfRangeException</span><span class="ss">)</span><span class="w"> </span>{<span class="w"> </span>
<span class="w">            </span><span class="o">//</span><span class="w"> </span><span class="nv">Log</span><span class="w"> </span><span class="k">then</span>,
<span class="w">            </span><span class="nv">throw</span><span class="w"> </span><span class="nv">ex</span><span class="c1">;</span>
<span class="w">        </span>}
<span class="w">        </span><span class="k">if</span><span class="w"> </span><span class="ss">(</span><span class="nv">ex</span><span class="w"> </span><span class="nv">is</span><span class="w"> </span><span class="nv">InvalidOperationException</span><span class="ss">)</span><span class="w"> </span>{
<span class="w">            </span><span class="o">//</span><span class="w"> </span><span class="k">Show</span><span class="w"> </span><span class="nv">message</span><span class="w"> </span><span class="k">then</span>,
<span class="w">            </span><span class="nv">throw</span><span class="w"> </span><span class="nv">ex</span><span class="c1">;</span>
<span class="w">        </span>}
<span class="w">        </span><span class="o">//</span><span class="w"> </span><span class="nv">and</span><span class="w"> </span><span class="nv">so</span><span class="w"> </span><span class="nv">on</span>.
<span class="w">    </span>}
}
</pre></div>

<p>If <code>try &amp; catch</code> were used in the <code>Main</code>, then I would use <code>throw;</code> to rethrow
the error. But in the above simplied code, all exceptions go through
<code>HandleException</code></p>
<p>Does <code>throw ex;</code> has the same effect as calling <code>throw</code> when called inside
<code>HandleException</code>?</p>
<p><br><br></p>
<h2>Answer</h2>
<p>Yes, there is a difference.</p>
<ul>
<li>
<p><code>throw ex</code> resets the stack trace (so your errors would appear to originate from <code>HandleException</code>)</p>
</li>
<li>
<p><code>throw</code> doesn't - the original offender would be preserved.</p>
<div class="code"><pre class="code literal-block"> static void Main(string[] args)
</pre></div>

<p>{
     try
     {
         Method2();
     }
     catch (Exception ex)
     {
         Console.Write(ex.StackTrace.ToString());
         Console.ReadKey();
     }
 }</p>
<p>private static void Method2()
 {
     try
     {
         Method1();
     }
     catch (Exception ex)
     {
         //throw ex resets the stack trace Coming from Method 1 and propogates it to the caller(Main)
         throw ex;
     }
 }</p>
<p>private static void Method1()
 {
     try
     {
         throw new Exception("Inside Method1");
     }
     catch (Exception)
     {
         throw;
     }
 }</p>
</li>
</ul>
<p><br></p>
<h3>Suggest</h3>
<p>(I posted earlier, and @Marc Gravell has corrected me)</p>
<p>Here's a demonstration of the difference:</p>
<div class="code"><pre class="code literal-block"><span class="nt">static</span><span class="w"> </span><span class="nt">void</span><span class="w"> </span><span class="nt">Main</span><span class="o">(</span><span class="nt">string</span><span class="cp">[]</span><span class="w"> </span><span class="nt">args</span><span class="o">)</span><span class="w"> </span><span class="p">{</span>
<span class="w">    </span><span class="err">try</span><span class="w"> </span><span class="err">{</span>
<span class="w">        </span><span class="err">ThrowException1()</span><span class="p">;</span><span class="w"> </span><span class="err">//</span><span class="w"> </span><span class="err">line</span><span class="w"> </span><span class="err">19</span>
<span class="w">    </span><span class="p">}</span><span class="w"> </span><span class="nt">catch</span><span class="w"> </span><span class="o">(</span><span class="nt">Exception</span><span class="w"> </span><span class="nt">x</span><span class="o">)</span><span class="w"> </span><span class="p">{</span>
<span class="w">        </span><span class="err">Console.WriteLine("Exception</span><span class="w"> </span><span class="err">1:")</span><span class="p">;</span>
<span class="w">        </span><span class="err">Console.WriteLine(x.StackTrace)</span><span class="p">;</span>
<span class="w">    </span><span class="p">}</span>
<span class="w">    </span><span class="nt">try</span><span class="w"> </span><span class="p">{</span>
<span class="w">        </span><span class="err">ThrowException2()</span><span class="p">;</span><span class="w"> </span><span class="err">//</span><span class="w"> </span><span class="err">line</span><span class="w"> </span><span class="err">25</span>
<span class="w">    </span><span class="p">}</span><span class="w"> </span><span class="nt">catch</span><span class="w"> </span><span class="o">(</span><span class="nt">Exception</span><span class="w"> </span><span class="nt">x</span><span class="o">)</span><span class="w"> </span><span class="p">{</span>
<span class="w">        </span><span class="err">Console.WriteLine("Exception</span><span class="w"> </span><span class="err">2:")</span><span class="p">;</span>
<span class="w">        </span><span class="err">Console.WriteLine(x.StackTrace)</span><span class="p">;</span>
<span class="w">    </span><span class="p">}</span>
<span class="err">}</span>

<span class="nt">private</span><span class="w"> </span><span class="nt">static</span><span class="w"> </span><span class="nt">void</span><span class="w"> </span><span class="nt">ThrowException1</span><span class="o">()</span><span class="w"> </span><span class="p">{</span>
<span class="w">    </span><span class="err">try</span><span class="w"> </span><span class="err">{</span>
<span class="w">        </span><span class="err">DivByZero()</span><span class="p">;</span><span class="w"> </span><span class="err">//</span><span class="w"> </span><span class="err">line</span><span class="w"> </span><span class="err">34</span>
<span class="w">    </span><span class="p">}</span><span class="w"> </span><span class="nt">catch</span><span class="w"> </span><span class="p">{</span>
<span class="w">        </span><span class="err">throw</span><span class="p">;</span><span class="w"> </span><span class="err">//</span><span class="w"> </span><span class="err">line</span><span class="w"> </span><span class="err">36</span>
<span class="w">    </span><span class="p">}</span>
<span class="err">}</span>
<span class="nt">private</span><span class="w"> </span><span class="nt">static</span><span class="w"> </span><span class="nt">void</span><span class="w"> </span><span class="nt">ThrowException2</span><span class="o">()</span><span class="w"> </span><span class="p">{</span>
<span class="w">    </span><span class="err">try</span><span class="w"> </span><span class="err">{</span>
<span class="w">        </span><span class="err">DivByZero()</span><span class="p">;</span><span class="w"> </span><span class="err">//</span><span class="w"> </span><span class="err">line</span><span class="w"> </span><span class="err">41</span>
<span class="w">    </span><span class="p">}</span><span class="w"> </span><span class="nt">catch</span><span class="w"> </span><span class="o">(</span><span class="nt">Exception</span><span class="w"> </span><span class="nt">ex</span><span class="o">)</span><span class="w"> </span><span class="p">{</span>
<span class="w">        </span><span class="err">throw</span><span class="w"> </span><span class="err">ex</span><span class="p">;</span><span class="w"> </span><span class="err">//</span><span class="w"> </span><span class="err">line</span><span class="w"> </span><span class="err">43</span>
<span class="w">    </span><span class="p">}</span>
<span class="err">}</span>

<span class="nt">private</span><span class="w"> </span><span class="nt">static</span><span class="w"> </span><span class="nt">void</span><span class="w"> </span><span class="nt">DivByZero</span><span class="o">()</span><span class="w"> </span><span class="p">{</span>
<span class="w">    </span><span class="err">int</span><span class="w"> </span><span class="err">x</span><span class="w"> </span><span class="err">=</span><span class="w"> </span><span class="err">0</span><span class="p">;</span>
<span class="w">    </span><span class="err">int</span><span class="w"> </span><span class="err">y</span><span class="w"> </span><span class="err">=</span><span class="w"> </span><span class="err">1</span><span class="w"> </span><span class="err">/</span><span class="w"> </span><span class="err">x</span><span class="p">;</span><span class="w"> </span><span class="err">//</span><span class="w"> </span><span class="err">line</span><span class="w"> </span><span class="err">49</span>
<span class="p">}</span>
</pre></div>

<p>and here is the output:</p>
<div class="code"><pre class="code literal-block">Exception 1:
   at UnitTester.Program.DivByZero() in &lt;snip&gt;\Dev\UnitTester\Program.cs:line 49
   at UnitTester.Program.ThrowException1() in &lt;snip&gt;\Dev\UnitTester\Program.cs:line 36
   at UnitTester.Program.TestExceptions() in &lt;snip&gt;\Dev\UnitTester\Program.cs:line 19

Exception 2:
   at UnitTester.Program.ThrowException2() in &lt;snip&gt;\Dev\UnitTester\Program.cs:line 43
   at UnitTester.Program.TestExceptions() in &lt;snip&gt;\Dev\UnitTester\Program.cs:line 25
</pre></div>

<p>You can see that in Exception 1, the stack trace goes back to the
<code>DivByZero()</code> method, whereas in Exception 2 it does not.</p>
<p>Take note, though, that the line number shown in <code>ThrowException1()</code> and
<code>ThrowException2()</code> is the line number of the <code>throw</code> statement, <strong>not</strong> the
line number of the call to <code>DivByZero()</code>, which probably makes sense now that
I think about it a bit...</p>
<h4>Output in Release mode</h4>
<p>Exception 1:</p>
<div class="code"><pre class="code literal-block">at ConsoleAppBasics.Program.ThrowException1()
at ConsoleAppBasics.Program.Main(String[] args)
</pre></div>

<p>Exception 2:</p>
<div class="code"><pre class="code literal-block">at ConsoleAppBasics.Program.ThrowException2()
at ConsoleAppBasics.Program.Main(String[] args)
</pre></div>

<p>Is it maintains the original stackTrace in debug mode only?</p>
    </div>
    <aside class="postpromonav"><nav><ul itemprop="keywords" class="tags">
<li><a class="tag p-category" href="../../categories/net/" rel="tag">.net</a></li>
            <li><a class="tag p-category" href="../../categories/net-c/" rel="tag">.net-c#</a></li>
            <li><a class="tag p-category" href="../../categories/exception/" rel="tag">exception</a></li>
        </ul>
<ul class="pager hidden-print">
<li class="previous">
                <a href="../how-to-uncheck-a-radio-button/" rel="prev" title="How to uncheck a radio button?">Previous post</a>
            </li>
            <li class="next">
                <a href="../how-can-i-post-data-as-form-data-instead-of-a-request-payload/" rel="next" title="How can I post data as form data instead of a request payload?">Next post</a>
            </li>
        </ul></nav></aside></article>
</div>
        <!--End of body content-->

        <footer id="footer">
            Contents © 2023         Go to StackOverflow Chinese Site  <a href="http://stackoverflow.ink">StackOverflow-ZH</a>  
            
        </footer>
</div>
</div>


            <script src="../../assets/js/all-nocdn.js"></script><!-- fancy dates --><script>
    moment.locale("en");
    fancydates(0, "YYYY-MM-DD HH:mm");
    </script><!-- end fancy dates --><script>
    baguetteBox.run('div#content', {
        ignoreClass: 'islink',
        captions: function(element) {
            return element.getElementsByTagName('img')[0].alt;
    }});
    </script><script src="../../assets/js/search.js"></script>
</body>
</html>
